sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ToNumberParser.scala

```diff
@@ -599,4 +614,254 @@ class ToNumberParser(numberFormat: String, errorOnFail: Boolean) extends Seriali
+          }
+      }
+    }
+
+    if (formattingBeforeDecimalPointIndex < inputBeforeDecimalPoint.length ||
+      formattingAfterDecimalPointIndex < inputAfterDecimalPoint.length) {
+      // Remaining digits before or after the decimal point exist in the decimal value but not in
+      // the format string.
+      formatMatchFailure(input, numberFormat)
+    } else {
+      stripTrailingLoneDecimalPoint(result)
+      if (result.isEmpty || result.toString == "+" || result.toString == "-") {
+        result.clear()
+        result.append('0')
+      }
+      UTF8String.fromString(result.toString())
+    }
+  }
+
+  /**
+   * Splits the provided Decimal value's string representation by the decimal point, if any.
+   * @param input the Decimal value to consume
+   * @return two strings representing the contents before and after the decimal point (if any)
+   */
+  private def formatSplitInputBeforeAndAfterDecimalPoint(input: Decimal): (String, String) = {
+    // Convert the input Decimal value to a string (without exponent notation).
+    val inputString = input.toJavaBigDecimal.toPlainString
+    // Split the digits before and after the decimal point.
+    val tokens: Array[String] = inputString.split(POINT_SIGN)
+    var beforeDecimalPoint: String = tokens(0)
+    var afterDecimalPoint: String = if (tokens.length > 1) tokens(1) else ""
+    // Strip any leading minus sign to consider the digits only.
+    // Strip leading and trailing zeros to match cases when the format string begins with a decimal
+    // point.
+    beforeDecimalPoint = beforeDecimalPoint.dropWhile(c => c == MINUS_SIGN || c == ZERO_DIGIT)
+    afterDecimalPoint = afterDecimalPoint.reverse.dropWhile(_ == ZERO_DIGIT).reverse
+
+    // If the format string specifies more digits than the 'beforeDecimalPoint', prepend leading
+    // spaces to make them the same length. Likewise, if the format string specifies more digits
+    // than the 'afterDecimalPoint', append trailing spaces to make them the same length. This step
+    // simplifies logic consuming the format tokens later.
+    var reachedDecimalPoint = false
+    var numFormatDigitsBeforeDecimalPoint = 0
+    var numFormatDigitsAfterDecimalPoint = 0
+    formatTokens.foreach {
+      case digitGroups: DigitGroups =>
+        digitGroups.digits.foreach { digits =>
+          val numDigits = digits match {
+            case ExactlyAsManyDigits(num) => num
+            case AtMostAsManyDigits(num) => num
+          }
+          for (_ <- 0 until numDigits) {
+            if (!reachedDecimalPoint) {
+              numFormatDigitsBeforeDecimalPoint += 1
+            } else {
+              numFormatDigitsAfterDecimalPoint += 1
+            }
+          }
+        }
+      case _: DecimalPoint =>
+        reachedDecimalPoint = true
+      case _ =>
+    }
+    // If there were more digits in the provided input string (before or after the decimal point)
+    // than specified in the format string, this is an overflow.
```

Comment:
where do we fail for the overflow case?

Reply:
The function evaluation does not return an error, but instead converts all the digits to `#`:

<img width="1874" alt="image" src="https://user-images.githubusercontent.com/99207096/167676555-cdf442e5-4c98-47bb-90c0-a2cdf0b340a1.png">

This is consistent with what e.g. Postgres and Oracle do.